torch/distributed/nn/api/remote_module.py

```diff
@@ -214,105 +209,155 @@ def __init__(
 
-                # TODO: We need to change this to rpc.remote, and make it async (see the else branch below).
-                # For that we need to be able to apply _module_interface_cls to the RRef returned by rpc.remote
-                # See https://github.com/pytorch/pytorch/issues/58098 for more context.
-                self.module_rref = rpc.rpc_sync(
-                    self.on,
-                    _create_module_with_interface,
-                    (module_cls, args, kwargs, self.device, _module_interface_cls),
-                )
-            else:
-                self.module_rref = module_rref
+            # Create the module on the remote side.
+            fut.wait()  # Ensure remote_module_cls is available on remote side.
+
+            # TODO: We need to change this to rpc.remote, and make it async (see the else branch below).
+            # For that we need to be able to apply _module_interface_cls to the RRef returned by rpc.remote
+            # See https://github.com/pytorch/pytorch/issues/58098 for more context.
+            self.module_rref = rpc.rpc_sync(
+                self.on,
+                _create_module_with_interface,
+                (module_cls, args, kwargs, self.device, _module_interface_cls),
+            )
         else:
             self.is_scriptable = False
             self.generated_methods = (
                 _NON_SCRIPTABLE_REMOTE_MODULE_MODULE._generated_methods
             )
-            if module_rref is None:
-                # Create the module on the remote side.
-                self.module_rref = rpc.remote(
-                    self.on,
-                    _create_module,
-                    (module_cls, args, kwargs, self.device),
-                )
-            else:
-                self.module_rref = module_rref
+            # Create the module on the remote side.
+            self.module_rref = rpc.remote(
+                self.on,
+                _create_module,
+                (module_cls, args, kwargs, self.device),
+            )
 
-        # Install generated methods.
-        for method in self.generated_methods:
-            method_name = method.__name__
-            method = torch.jit.export(method)
-            setattr(self, method_name, types.MethodType(method, self))
+        self._install_generated_methods()
+        self._check_attribute_picklability()
 
-        # Sanity check: whether to be pickled must be explicitly defined for every attribute.
-        for k in self.__dict__.keys():
-            if (
-                k not in _REMOTE_MODULE_PICKLED_ATTRIBUTES
-                and k not in _REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING
-            ):
-                raise AttributeError(
-                    "Attribute {} must be either in ``_REMOTE_MODULE_PICKLED_ATTRIBUTES`` or "
-                    "``_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING``.".format(k)
-                )
+    def init_from_module_rref(
+        self,
+        remote_device: str,
```

Comment:
Is this `remote_device` a necessary argument? from the example I see that we are duplicating the "worker1/cpu". I think `init_from_module_rref` is a API that tries to initialize RemoteModule locally on the current worker, and since we actually don't the real RemoteModule (the real one is the RRef[Module] passed in), so it seems we don't need to worry about the device that we need to place into?

Reply:
It's needed to infer `enable_moving_cpu_tensors_to_cuda` and hence affect `_init_template`.

Reply:
I see, I guess it's inevictable as RRef itself doesn't store information like this.